### DISCUSSION TITLE
Restructuring Bootstrap Service

### DISCUSSION THREAD

**Discussion by arvindh123** [absmach/magistrala#2244] [Ideas]
Bootstrap repo contains `things` table, `channels` table and `thing channel connections` table, i.e  table will contains things and channels many to many relationship.   
 
Actually in Magistrala 0.14 `thing channel connections` table is remove and their relations are stored in policy engine.

Still we need `thing channel connections` table in bootstrap. Can we remove this table? 

What the effects will their if we remove `thing channel connections` table .

**Comment by arvindh123** [comment]
@arvindh123 
> When a device get bootstrap config, Then the device is intended to use that thing ID in bootstrap config.
> The device intended to send messages to Magistrala via the channels in bootstrap.
> 
> Lets take example:
> If remote device get the bootstrap config, Then device will use the thingID and it will use two channels.`channel1` for temperature data and `channel2` for humidity data
> 
> If that remote device's thingID is disconnected from any one of the channel manually
> 
> Then , If `channel1` is disconnected then ,  Device could not send data to Magistala via `channel1` 
> or
> If `channel2` is disconnected then, Device could not send data to Magistala via `channel2` 
> So the device is not having Active bootstrap config. We can say it is in partially active state
> since we don't have third state now, simply we can say it is inactive 


@dborovcanin 

> This is true for disconnect. However, if we reconnect - we can't know if all the channels that are supposed to be connected are actually connected. I.e. the approach we have works for one channel only. If we disconnect from two channels and connected to one - it will set status to active

@arvindh123 
> Yeah I agree, But this also we solve,
> In Bootstrap config ,we have have all thing related Channel IDs ,
> So if thing got reconnect to channel,
> we can check bootstrap thing is connected to all of its channels in mentioned in bootstrap config
> and
> then we can decide the state of bootstrap.

@dborovcanin 

> Yes, that's what we should do. We should do that anyways since not all the Channels are relevant for all the configs, while that's unlikely.



From the above conversations, I can say, we need things and channels connection list in bootstrap in order to verify the state of config. 

So lets have the things and channels connection table in bootstrap service 

And Upgrade the bootstrap thing connect event handler.

Spec: 
if thing got reconnect to channel, then check bootstrap thing is connected to all of its channels in mentioned in bootstrap config and then decide the state of bootstrap config is `active` or `inactive` or intermediate state.

**Comment by arvindh123** [comment]
https://github.com/absmach/magistrala/pull/2192#issuecomment-2148960908

https://github.com/absmach/magistrala/pull/2192#issuecomment-2157606840

From the above comments, 
We need to decide whether to have `intermediate state` beyond  `active` and `inactive`.

**Reply by JeffMboya** [reply]
I think we need an intermediate state
